src/clj/rems/db/applications.clj

```diff
@@ -1,31 +1,284 @@
+  (-> event
+      coerce-event-commons
+      coerce-event-specifics))
+
+(defn json->event [json]
+  (when json
+    (coerce-event (json/parse-string json))))
+
+(defn event->json [event]
+  (events/validate-event event)
+  (json/generate-string event))
+
+(defn- fix-event-from-db [event]
+  (assoc (-> event :eventdata json->event)
+         :event/id (:id event)))
+
+(defn get-application-events [application-id]
+  (assert application-id)
+  (->> {:application application-id}
+       db/get-application-events
+       (mapv fix-event-from-db)))
+
+(defn get-all-events-since [event-id]
+  (mapv fix-event-from-db (db/get-application-events-since {:id event-id})))
+
+(defn get-latest-event []
+  (fix-event-from-db (db/get-latest-application-event {})))
+
+(defn- user-applications-projection [user-applications event application]
+  (let [actor (:event/actor event)
+        app-id (:application/id event)]
+    (case (:event/type event)
+      (:application.event/attachments-redacted
+       :application.event/applicant-changed
+       :application.event/approved
+       :application.event/closed
+       :application.event/review-requested
+       :application.event/reviewed
+       :application.event/copied-to
+       :application.event/decided
+       :application.event/decider-invited
+       :application.event/decision-requested
+       :application.event/deleted
+       :application.event/draft-saved
+       :application.event/external-id-assigned
+       :application.event/expiration-notifications-sent
+       :application.event/licenses-accepted
+       :application.event/licenses-added
+       :application.event/member-invited
+       :application.event/member-uninvited
+       :application.event/rejected
+       :application.event/remarked
+       :application.event/resources-changed
+       :application.event/returned
+       :application.event/reviewer-invited
+       :application.event/revoked
+       :application.event/submitted
+       :application.event/voted)
+      user-applications ; no user added
+
+      :application.event/copied-from
+      (reduce (fn [user-applications userid]
+                (update user-applications userid app-id))
+              user-applications
+              (:application/members application))
+
+      :application.event/created
+      (update user-applications actor conj-set app-id)
+
+      :application.event/decider-joined
+      (update user-applications actor conj-set app-id)
+
+      :application.event/member-added
+      (update user-applications actor conj-set (:userid (:application/member event)))
+
+      :application.event/member-joined
+      (update user-applications actor conj-set app-id)
+
+      :application.event/member-removed
+      (update user-applications actor disj-set (:userid (:application/member event)))
+
+      :application.event/reviewer-joined
+      (update user-applications actor conj-set app-id))))
+
+(defn- user-roles-projection [user-roles event application]
```

Comment:
Could this name be `user-application-roles-projection` or so because these are only the roles that come from applications. Not workflow handlers, roles table etc.

Reply:
Consider also just listening to `application` and updating based on its `:application/user-roles` data that is enriched.